### PR TITLE
[clang-tidy] fix false positive in bugprone-return-const-ref-from-parameter

### LINKED ISSUE #115743: [clang-tidy] False-positive with returning a constant reference parameter.
Hi everyone,

Consider the following snippet of code:

```cpp
#include <array>

template<class T>
void f(const T& t) {
    const auto get = [&t] -> const T& { return t; };
    return std::array{get(), get(), get()};
}
```

When I run the latest released `clang-tidy` (19.1.3) over it with a check `bugprone-return-const-ref-from-parameter`, it produces a false-positive warning:

```
$ clang-tidy --version                                                                                                                              
Homebrew LLVM version 19.1.3
  Optimized build.
$ clang-tidy b.cpp "--checks=bugprone-return-const-ref-from-parameter" -- -std=c++23                                                      
1 warning generated.
.../b.cpp:5:45: warning: returning a constant reference parameter may cause use-after-free when the parameter is constructed from a temporary [bugprone-return-const-ref-from-parameter]
    5 |         const auto get = [&t] -> const T& { return t; };
      |                                      
```

Clearly, `t` is not returned from the function `f`, and it is not a parameter of the lambda.

### PR DESCRIPTION
**Pull request by jcsxky**
Fix https://github.com/llvm/llvm-project/issues/115743

### CODE REVIEW

**Comment by EugeneZelenko** [comment]
Please rebase from current `main`.

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -179,7 +179,8 @@ Changes in existing checks
 - Improved :doc:`bugprone-return-const-ref-from-parameter
   <clang-tidy/checks/bugprone/return-const-ref-from-parameter>` check to
   diagnose potential dangling references when returning a ``const &`` parameter
-  by using the conditional operator ``cond ? var1 : var2``.
+  by using the conditional operator ``cond ? var1 : var2``and no longer giving
```
Comment (EugeneZelenko) [inline clang-tools-extra/docs/ReleaseNotes.rst:182]:
```suggestion
  by using the conditional operator ``cond ? var1 : var2`` and no longer giving
```

File: clang-tools-extra/test/clang-tidy/checkers/bugprone/return-const-ref-from-parameter.cpp [hunk 2]
```diff
@@ -151,6 +151,12 @@ void instantiate(const int &param, const float &paramf, int &mut_param, float &m
         itf6(mut_paramf);
 }
 
+template<class T>
+void f(const T& t) {
+    const auto get = [&t] -> const T& { return t; };
+    return T{};
+}
```
Comment (PiotrZSL) [inline clang-tools-extra/test/clang-tidy/checkers/bugprone/return-const-ref-from-parameter.cpp:158]: Please add test:
```
const auto get = [](const T& t2) -> const T& { return t2; };
```
and verify that it's being still detected.

File: clang-tools-extra/clang-tidy/bugprone/ReturnConstRefFromParameterCheck.cpp [hunk 3]
```diff
@@ -18,14 +18,17 @@ namespace clang::tidy::bugprone {
...
-          to(parmVarDecl(hasType(hasCanonicalType(
-                             qualType(lValueReferenceType(pointee(
-                                          qualType(isConstQualified()))))
-                                 .bind("type"))))
+          to(parmVarDecl(
+                 hasType(hasCanonicalType(
+                     qualType(lValueReferenceType(
+                                  pointee(qualType(isConstQualified()))))
+                         .bind("type"))),
+                 parmVarDecl(hasDeclContext(functionDecl().bind("owner"))))
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/ReturnConstRefFromParameterCheck.cpp:26]: no need for duplicating parmVarDecl

**Review by PiotrZSL — CHANGES_REQUESTED** [review]
Please add positive test for lambdas.

**Comment by jcsxky** [comment]
> Please add positive test for lambdas.

Positive test cases have been added.

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 4]
```diff
@@ -179,7 +179,8 @@ Changes in existing checks
 - Improved :doc:`bugprone-return-const-ref-from-parameter
   <clang-tidy/checks/bugprone/return-const-ref-from-parameter>` check to
   diagnose potential dangling references when returning a ``const &`` parameter
-  by using the conditional operator ``cond ? var1 : var2``.
+  by using the conditional operator ``cond ? var1 : var2`` and no longer giving
```
Comment (HerrCai0907) [inline clang-tools-extra/docs/ReleaseNotes.rst:182]: I am not sure it is a fix of false positives. Maybe in generic, it more looks like add support of lambda function.

**Review by HerrCai0907 — APPROVED** [review]

**Pull request merged**